### PR TITLE
Refactored the GeoBoundsAggregation code from server folder to geo module.

### LINKED ISSUE #3980: Implementation of GeoBounds Aggregation on Geo Shape field
### Description
This issue tracks the implementation of GeoBounds Aggregation on Geo Shape field. 

### Tasks:
- [x] Refactor code of GeoBounds Aggregation on GeoPoint from server to modules/geo folder.
- [x] Backport the refactor code to 2.x branch
- [x] Implement the GeoBounds on GeoShape field
     - [x] Implement the code to convert the geobounds returned from doc value to internal aggregation.
     - [x] Integration Tests
     - [x] Unit Tests

### Dependency
https://github.com/opensearch-project/OpenSearch/issues/4049

### Useful Links
1. RFC for All Aggregations on GeoShape: https://github.com/opensearch-project/geospatial/issues/84
2. GeoBounds RFC on GeoShape : https://github.com/opensearch-project/geospatial/issues/93
3. Low Level Implementation Proposal: https://github.com/opensearch-project/geospatial/issues/92
4. https://github.com/opensearch-project/OpenSearch/issues/4049
5. All Aggregation Implementation issue: https://github.com/opensearch-project/geospatial/issues/104

### PR DESCRIPTION
**Pull request by navneet1v**
### Description
This change refactors the code of GeoBoundsAggregation on GeoPoints from the server folder to the modules/geo. 

The proposal is being discussed and signed off here: https://github.com/opensearch-project/geospatial/issues/92
 
### Issues Resolved
#3980

High Level Task: https://github.com/opensearch-project/geospatial/issues/104
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Signed-off-by: Navneet Verma <navneev@amazon.com>

### CODE REVIEW

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/GeoPluginIntegTestCase.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+     * This was added as a backdoor to Mock the implementation of {@link GeoShapeFieldMapper} which was coming from
+     * {@link GeoPlugin}. Mock implementation is {@link TestGeoShapeFieldMapperPlugin}. Now we are using the
+     * {@link GeoPlugin} in our integration tests we need to override this functionality to avoid multiple mapper
+     * error.
+     *
+     * @return boolean
+     */
+    @Override
+    protected boolean addMockGeoShapeFieldMapper() {
+        return false;
```
Comment (nknize) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/GeoPluginIntegTestCase.java:45]: In what test scenarios would this get set to `true`?

File: modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/GeoBoundsAggregator.java [hunk 2]
```diff
@@ -51,22 +47,11 @@
...
-
-    private final ValuesSource.GeoPoint valuesSource;
-    private final boolean wrapLongitude;
-    DoubleArray tops;
-    DoubleArray bottoms;
-    DoubleArray posLefts;
-    DoubleArray posRights;
-    DoubleArray negLefts;
-    DoubleArray negRights;
+final class GeoBoundsAggregator extends AbstractGeoBoundsAggregator<ValuesSource.GeoPoint> {
```
Comment (nknize) [inline modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/GeoBoundsAggregator.java:54]: Will there be a new `ValuesSource.GeoShape` when shape docvalues are added to lucene? Or are you planning on creating a `ValuesSource.Geo` that handles both?

**Review by nknize — COMMENTED** [review]
Love this! I have just a couple initial questions before fully approving.

File: modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/GeoBoundsAggregator.java (hunk 2 above)
Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/GeoBoundsAggregator.java:54]: I planning to create a new ValueSource which will be ValueSource.GeoShape for geoshapes. Now I think about it, may be we can combine them form and single coresource value type. I might need to check.

But for this PR I don't want to change it. I think I will take it in another PR as this PR is getting long already.

I will do it in the followup PR, as there we will be a lot of classes that will be required to be changed.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/GeoPluginIntegTestCase.java (hunk 1 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/GeoPluginIntegTestCase.java:45]: As per the documentation there are test cases which are written in the server folder which depends on this, which test the indexing of geoshape as the geoshape mapper is still in the server folder. But on running this code without this makes the result successful. Let me just update the PR by removing these functions.

**Comment by VijayanB** [comment]
@navneet1v Since you are refactoring/moving one aggregation at a time, shall we create tasks for list of aggregation that will be refactored?

**Comment by navneet1v** [comment]
> @navneet1v Since you are refactoring/moving one aggregation at a time, shall we create tasks for list of aggregation that will be refactored and link to corresponding PR?

Sure. I think I had this thing in mind but I forgot. Let me create an issue and link it in all the places.

The issue will be an uber issue which will list RFCs, and implementation issues for every Aggregation.

**Comment by navneet1v** [comment]
> > @navneet1v Since you are refactoring/moving one aggregation at a time, shall we create tasks for list of aggregation that will be refactored and link to corresponding PR?
> 
> Sure. I think I had this thing in mind but I forgot. Let me create an issue and link it in all the places.
> 
> The issue will be an uber issue which will list RFCs, and implementation issues for every Aggregation.

Issue: https://github.com/opensearch-project/geospatial/issues/104

File: modules/geo/build.gradle [hunk 3]
```diff
@@ -28,9 +28,10 @@
  * under the License.
  */
 apply plugin: 'opensearch.yaml-rest-test'
+apply plugin: 'opensearch.internal-cluster-test'
 
 opensearchplugin {
-  description 'Placeholder plugin for geospatial features in OpenSearch. only registers geo_shape field mapper for now'
+  description 'Plugin for geospatial features in OpenSearch. Registering the geo_shape and aggregations GeoBounds on Geo_Shape and Geo_Point'
```
Comment (VijayanB) [inline modules/geo/build.gradle:34]: Do we need to add this to description?
```suggestion
  description 'Plugin for geospatial features in OpenSearch.'
```

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/metrics/GeoPluginBoundsIT.java [hunk 4]
```diff
@@ -43,21 +43,21 @@
...
+import static org.hamcrest.Matchers.closeTo;
+import static org.opensearch.index.query.QueryBuilders.matchAllQuery;
+import static org.opensearch.search.aggregations.AggregationBuilders.global;
+import static org.opensearch.search.aggregations.AggregationBuilders.terms;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertSearchResponse;
+import static org.opensearch.geo.tests.common.AggregationBuilders.geoBounds;
 
 @OpenSearchIntegTestCase.SuiteScopeTestCase
-public class GeoBoundsIT extends AbstractGeoTestCase {
+public class GeoPluginBoundsIT extends AbstractGeoAggregatorTestCasePlugin {
```
Comment (VijayanB) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/metrics/GeoPluginBoundsIT.java:60]: Aggregation name is geo bounds. Why are we call this as geo plugin bounds?

File: modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoBoundsAggregator.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+import org.opensearch.search.aggregations.metrics.MetricsAggregator;
+import org.opensearch.search.aggregations.support.ValuesSource;
+import org.opensearch.search.aggregations.support.ValuesSourceConfig;
+import org.opensearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Abstract class for doing the {@link GeoBounds} Aggregation over fields of type GeoShape and GeoPoint.
```
Comment (VijayanB) [inline modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoBoundsAggregator.java:25]:
```suggestion
 * Abstract class for doing the {@link GeoBounds} Aggregation over fields of type geo_shape and geo_point.
```

Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoBoundsAggregator.java:25]: what is the benefit for geo_shape and geo_point here?

File: modules/geo/build.gradle (hunk 3 above)
Comment (navneet1v) [reply modules/geo/build.gradle:34]: I just updated the description as it was not correct. I want to provide exact information what would be provided from this plugin. "Plugin for geospatial features in OpenSearch." This statement is not 100% correct all alone. Hence added what other things will be added from this plugin for now.

Once we keep on moving more things and from Geo standpoint, everything becomes consolidated we can write it provides GeoSpatial feature.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/metrics/GeoPluginBoundsIT.java (hunk 4 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/metrics/GeoPluginBoundsIT.java:60]: I guess this is some refactoring error from my side. I might be testing 2 IT tests at once. Will fix it.

File: modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoBoundsAggregator.java (hunk 5 above)
Comment (VijayanB) [reply modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoBoundsAggregator.java:25]: Sure, if it is not useful, we can remove it. Suggestion is if we are adding explicitly, lets call field type with right case.

Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoBoundsAggregator.java:25]: I will update it, as it is a nit-pick

File: modules/geo/src/main/java/org/opensearch/geo/GeoPlugin.java [hunk 6]
```diff
@@ -32,18 +32,36 @@
...
+     * Registering {@link GeoBounds} aggregation on GeoPoint field.
+     */
+    @Override
+    public List<AggregationSpec> getAggregations() {
+        final AggregationSpec spec = new AggregationSpec(
+            GeoBoundsAggregationBuilder.NAME,
+            GeoBoundsAggregationBuilder::new,
+            GeoBoundsAggregationBuilder.PARSER
+        ).addResultReader(InternalGeoBounds::new).setAggregatorRegistrar(GeoBoundsAggregationBuilder::registerAggregators);
+        return Collections.singletonList(spec);
```
Comment (VijayanB) [inline modules/geo/src/main/java/org/opensearch/geo/GeoPlugin.java:65]: nit
```suggestion
        return List.of(spec);
```

**Review by VijayanB — APPROVED** [review]

File: modules/geo/src/main/java/org/opensearch/geo/GeoPlugin.java (hunk 6 above)
Comment (navneet1v) [reply modules/geo/src/main/java/org/opensearch/geo/GeoPlugin.java:65]: what is the difference between both of them? I have seen Collections.singletonList used a lot of time in the code, so I thought to use this.

**Comment by navneet1v** [comment]
@opensearch-project/opensearch-core can anyone take a look into this CR, it has been there for long period of time.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoAggregatorTestCaseModulePlugin.java [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertSearchResponse;
+
+/**
+ * This is base class for all Geo Aggregations Integration Tests. This class is similar to what we have in the server
+ * folder of the OpenSearch repo. As part of moving the Geo based aggregation into separate module and plugin we need
+ * to copy the code as we cannot depend on this class.
+ * <a href="https://github.com/opensearch-project/geospatial/issues/92">GitHub issue</a>
+ */
+public abstract class AbstractGeoAggregatorTestCaseModulePlugin extends GeoModulePluginIntegTestCase {
```
Comment (nknize) [inline modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoAggregatorTestCaseModulePlugin.java:45]: nit: in a follow on PR let's refactor this to `AbstractGeoAggregatorModulePluginTestCase`.

**Review by nknize — APPROVED** [review]
LGTM! One minor nit we can do in a follow up PR.

**Comment by navneet1v** [comment]
> LGTM! One minor nit we can do in a follow up PR.

sure will fix in the next PR for moving the GeoTile and GeoHashGird aggregation to Geo module

**Comment by navneet1v** [comment]
Backport of 2.x is pending. Need to figure out a way to backport these changes to the 2.x branch.

File: modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoAggregatorTestCaseModulePlugin.java (hunk 7 above)
Comment (navneet1v) [reply modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/metrics/AbstractGeoAggregatorTestCaseModulePlugin.java:45]: PR: https://github.com/opensearch-project/OpenSearch/pull/4180 with this change

**Comment by navneet1v** [comment]
> The backport to `2.x` failed:
> 
> ```
> The process '/usr/bin/git' failed with exit code 1
> ```
> 
> To backport manually, run these commands in your terminal:
> 
> [BASH (SHELL) (16 lines)](#)
> ```shell
> # Fetch latest updates from GitHub
> git fetch
> # Create a new working tree
> git worktree add .worktrees/backport-2.x 2.x
> # Navigate to the new working tree
> cd .worktrees/backport-2.x
> # Create a new branch
> git switch --create backport/backport-3992-to-2.x
> # Cherry-pick the merged commit of this pull request and resolve the conflicts
> git cherry-pick -x --mainline 1 4751f3bd296e6952b852a457a82220ec248b037a
> # Push it to GitHub
> git push --set-upstream origin backport/backport-3992-to-2.x
> # Go back to the original working tree
> cd ../..
> # Delete the working tree
> git worktree remove .worktrees/backport-2.x
> ```
> 
> Then, create a pull request where the `base` branch is `2.x` and the `compare`/`head` branch is `backport/backport-3992-to-2.x`.

Back-porting PR: https://github.com/opensearch-project/OpenSearch/pull/4179

**Pull request merged**